### DISCUSSION TITLE
Resizing Image control costs too much memory

### DISCUSSION THREAD

**Discussion by Supheria** [AvaloniaUI/Avalonia#17660] [Q&A] [answered]
```xaml
<UserControl
    ...>
    <Image Source="{Binding Source}" />
</UserControl>

```

------

```c#
public partial class Thumbnail : UserControl
![screenshot](https://github.com/user-attachments/assets/393d9c91-5571-443b-8122-dad944ca29ec)

{
    ...
        
    public Bitmap? Source
    {
        get => GetValue(SourceProperty);
        set => SetValue(SourceProperty, value);
    }
    public static readonly StyledProperty<Bitmap?> SourceProperty =
        AvaloniaProperty.Register<Thumbnail, Bitmap?>(nameof(Source));

    public Thumbnail()
    {
        InitializeComponent();
        AffectsRender<Thumbnail>(SourceProperty);
    }

    ...
}

```

------

```C#
internal partial class ThumbnailViewModel : ViewModelBase
{
    ...

    [ObservableProperty]
    Bitmap? _source = null;
    
    ...
}
```

------

As shown in code blocks, when Thumbnail.Source sets to a 2000*2000 bitmap, resizing Thumbnail will cause Image to resize too, which will cause a large cost of memery suddenly and dropdown, shown as screenshot below:
![screenshot](https://github.com/user-attachments/assets/c2f30c88-9f1e-4536-9b87-00bf12712bb4)

And if cost of memery dosen't meet to "maximum", memery cost will last for a long time with no disposing, shown as blow:

![screenshot2](https://github.com/user-attachments/assets/3b251f41-dcb9-411a-8dc0-b7c6aab7f086)
![screenshot3](https://github.com/user-attachments/assets/eec59337-c6b6-49cb-937f-35ae90459c65)

**How could I avoid this condition, or dispose memory cost in time in Render of Image？**

**Comment by timunie** [comment]
2000x2000 is no thumbnail imo. Consider to use Bitmap.DecodeToWidth and limit the size to whatever suits you.

**Reply by Supheria** [reply]
> 2000x2000 is no thumbnail imo. Consider to use Bitmap.DecodeToWidth and limit the size to whatever suits you.

It seems working nice for thumbnail, however thumbnail here is just for example. I'm not sure whether Image can dispose Source properly or not while Souce changed. 

In case if I use Bitmap as user interface to interact with mouse pointer, Source of Image may change frequently when mouse dragging. I hope Bitmap can dispose its data automatically by GC. And if such a UI sizes to fullscreen, resizing Source may also a problem for Image, I hope Bitmap.DecodeToWidth  works well as for thumbnail.

By the way, my thumbnail here is generated by WriteableBitmap, is there any way simple converting it to Stream that using in Bitmap.DecodeToWidth?

**Reply by maxkatz6** [reply, marked as answer]
> I hope Bitmap can dispose its data automatically by GC

It does.

> By the way, my thumbnail here is generated by WriteableBitmap, is there any way simple converting it to Stream that using in Bitmap.DecodeToWidth?

If you want to decode image, you need to encode it first. WriteableBitmap.Save for example.

**Reply by thevortexcloud** [reply]
>  Consider to use Bitmap.DecodeToWidth

Not sure if it has been fixed or not but based on my testing that still uses a lot of RAM. This however has all been discussed before. EG 

https://github.com/AvaloniaUI/Avalonia/discussions/13605
https://github.com/AvaloniaUI/Avalonia/issues/14304

**Reply by Supheria** [reply]
> > I hope Bitmap can dispose its data automatically by GC
> 
> It does.
> 
> > By the way, my thumbnail here is generated by WriteableBitmap, is there any way simple converting it to Stream that using in Bitmap.DecodeToWidth?
> 
> If you want to decode image, you need to encode it first. WriteableBitmap.Save for example.

Alright, I'll try to use WriteableBitmap.Save to save it as stream then use DecodeToWidth, since WriteableBitmap.CreateScaledBitmap doesn't work now.

**Reply by Supheria** [reply]
> > Consider to use Bitmap.DecodeToWidth
> 
> Not sure if it has been fixed or not but based on my testing that still uses a lot of RAM. This however has all been discussed before. EG
> 
> #13605 #14304

Thanks. Problem I met now is just a single Image control changing or resizing its Source frequently. Bitmap can auto Dispose making it much easier.
